### ISSUE TITLE
Relation hydration fails, stays integer instead of creating Proxy

### ISSUE THREAD

**Issue by JKapitein** [doctrine/doctrine2#7140]
In specific cases in a project I've inherited, an entity relationship can fail to hydrate. So far, we've been able to use stopgap solutions, mostly applying eager fetching, but this bug keeps popping up in other places.

We were able to make a reproduction on a clean Symfony Flex project, which you can find here
https://github.com/JKapitein/doctrine-hydration

Steps to reproduce:

1. Load in a single entity where the relations go (at least?) 2 deep. In the reproduction, we load it in through the Symfony ParamConverter
2. Within the controller action, call a repository function to load a list of the same type as the entity, joining the relations in the QueryBuilder. The list must include the entity that was already loaded.
3. The 2nd level relation will not get hydrated, and will only return the scalar ID, leading to errors if you try to fetch data from that relation

We're not certain if this behaviour stems from the relationship structure (as the ID on Person is a relationship), or if there's any other reason, but guidance would be appreciated.

**Comment by Ocramius** [comment]
This requires an isolated test case to be added to https://github.com/doctrine/doctrine2/tree/41292e85dfadac771713882663ab5389e1070ecc/tests/Doctrine/Tests/ORM/Functional/Ticket

We can't do much with an example app.

**Comment by JKapitein** [comment]
It passes on master but not on the 2.7 branch. I've created the pull request against 2.7, since master seems like a rewrite of sorts?

**Comment by Ocramius** [comment]
@JKapitein yeah, `master` is `3.x` now. If you are fixing an issue affecting `2.x`, I suggest you open a PR against `2.6`

**Comment by JKapitein** [comment]
Thank you @Ocramius, I've recreated the pull request against 2.6

Best I can tell is that the 'fetched' hint in UnitOfWork::createEntity gets set to true for the "account" relation, presumably because of the query builder, so the proxy is never created for the already existing entity?
